rir/src/compiler/pir/instruction.cpp

```diff
@@ -87,20 +92,107 @@ void Instruction::printGraphBranches(std::ostream& out, size_t bbId) const {
     assert(false);
 }
 
+void Instruction::printEnv(std::ostream& out, bool tty) const {
+    if (hasEnv()) {
+        if (tty) {
+            if (leaksEnv())
+                ConsoleColor::magenta(out);
+            else if (changesEnv())
+                ConsoleColor::red(out);
+            else
+                ConsoleColor::yellow(out);
+        }
+
+        env()->printRef(out);
+
+        if (tty)
+            ConsoleColor::clear(out);
+    }
+}
+
+void Instruction::printEffects(std::ostream& out, bool tty) const {
```

Comment:
I have an alternative proposal: what about printing every effect on every instruction, but somehow abbreviate and group them into categories, such that it fits in 2-3 characters. For example:

`Any` => `*`

`Force | Reflection | LeakArg | ExecuteCode` => `#` 
`Force | ExecuteCode` => `>`

`Warn | Error  | Visibility` => `e`
`Warn | Visibility` => `w`
`Visibility` => `v`

Reads/writes/leaks env are already in the color of the env, they don't have to be printed.

ChangesContext, TriggerDeopt are specific to certain instructions and never removed, so I would not print them either.

Reply:
I can abbreviate them. Although I still think I should print out all effects (except maybe changescontext, triggerdeopt, read/write/leak env), because I can imagine getting a really confusing bug because an instruction has a strange set of effects.